File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    answer = int(1e18)
    # 기다리는 사람 수보다 더 많이 검사 가능한 경우
    ## 시간 줄이기
    
    # 기다리는 사람 수보다 더 적게 검사 가능한 경우
    ## 시간 늘리기
    
    l = 1
    r = int(1e18)
    while l <= r:
        mid = (l + r) // 2
        
        cnt = 0
        for t in times:
            cnt += mid // t
        
        if cnt >= n:
            answer = min(answer, mid)
            r = mid - 1
        else:
            l = mid + 1
            
    
    return answer

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    static int receptions;
    public long solution(int n, int[] times) {
        Arrays.sort(times);
        long answer = Long.MAX_VALUE;
        receptions = times.length;
        long l = 1;
        long r = times[times.length - 1] * (long) n;
        while (l <= r) {
            long mid = (l + r) / 2L;
            long cnt = 0L;
            for (int i=0; i<receptions; i++) {
                cnt += mid / (long)times[i];
            }
            if (cnt >= n) {
                r = mid - 1;
                answer = Math.min(answer, mid);
            } else {
                l = mid + 1;
            }
        }
        return answer;
    }
}